### ISSUE TITLE
Kill idle nodes after interval

### ISSUE THREAD

**Issue by ryudice** [zalando/zalenium#52]
Is there any way to shutdown idle nodes that for some reason where not shutdown by the driver?

**Comment by diemol** [comment]
Right now, there are only manual ways, but stopping the container or shutting down the node via the API.
We thought about this early in the development but we could not find a good reason to kill the idle nodes.

Do you have a use case/scenario in mind?

**Comment by diemol** [comment]
I will close this due to inactivity, please feel free to re-open in case it is needed.

Thanks!

**Comment by diemol** [comment]
We found a use case for this, where we need to kill the container in order to get the video when a test is idle or fails in a nasty way. Reopening :)

**Comment by diemol** [comment]
I like the idea of the parameter, I am just trying to avoid it now because this would demand validations and more parameter handling, etc...

First I will explore the internal methods available in the Selenium proxy to see if something can tell us that the test is idle or it has crashed.

When no other option is found, I guess the a new parameter for this will be used.

Any other ideas you guys have are welcome, also PRs :)

**Comment by svenhornberg** [comment]
I use your grid with the following docker-compose configuration: https://github.com/zalando/zalenium/blob/master/docs/docker-compose.yaml.

A first try could be something like this:
I did a [shellscript](https://gist.github.com/svenhornberg/9b36e788a94acc5ac57891c3332927ed) which checks each 10 min for a node which runs a hour and stop it.
If you can format the timeoutput to a long it easier to parameterize.

**Comment by diemol** [comment]
Thanks for this script @svenhornberg!

Nevertheless, the approach we are taking is similar to the one used by the grid cloud providers (Sauce Labs, BrowserStack, etc...). Inside the node, we will monitor sessions that have been idle for more than 90 seconds (this value can be configurable via a test capability), when an idle session is detected, the node will be shutdown and the video copied. We will provide clear documentation on how to use this.

I am working on this now, and we should be able to release this at the end of this week.

**Comment by svenhornberg** [comment]
@diemol nice to know :)

**Comment by diemol** [comment]
Hi,

We just released a new version of the images with this improvement. 

The default idle timeout is 90 seconds, which can be configured via test capabilities. More details [here](https://github.com/zalando/zalenium/blob/master/docs/usage_examples.md#idle-timeout)

Whenever you can, please have a look and, if possible, give us some feedback.

**Comment by diemol** [comment]
This feature is already live, I'll close this issue due to inactivity.

Feel free to reopen in case something related does not work as expected.

**Comment by ketchoop** [comment]
Does it work with k8s? It doesn't seem to be working with k8s.

**Comment by mohitsh08** [comment]
Hi @diemol,

I am not creating new ticket so adding a comment on this one. 
This one is urgent and we have a project release so please help as soon as possible.

**_Issue:_** Idle nodes are not getting killed. idleTimeout passed in desired capabilities is 80 sec.

Test execution approach: We have a inhouse web application for automation testing which allows multiple user to trigger execution. I am running execution from 5 different user. Everytime a user trigger execution a testng test suite is created which have around 40 to 50 test cases and each test case takes around 10 to 12 mins to complete as it is E2E flow. 5 test suites are executing in parallel and each suites also have a parallel execution which have thread count 2.   

I am using latest zalenium that 3.141.59 (downloaded image 3 days ago from docker hub)and below docker version:

Client:
 Version:      17.06.1-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   874a737
 Built:        Thu Aug 17 22:48:20 2017
 OS/Arch:      windows/amd64

Server:
 Version:      17.06.1-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   874a737
 Built:        Thu Aug 17 22:54:55 2017
 OS/Arch:      linux/amd64
 Experimental: true

I have installed docker on Windows server 2016 which have 48 gb ram.
I have assigned 4 cpu and 24 gb ram to docker.

Below is the docker shell script to start zalenium:
docker run --rm -ti --name zalenium -p 4444:4444 -e ZALENIUM_EXTRA_JVM_PARAMS="-Dwebdriver.http.factory=apache" -v /var/run/docker.sock:/var/run/docker.sock -v /tmp/videos:/home/seluser/videos -e "WAIT_TIMEOUT=10m" --privileged dosel/zalenium start --desiredContainers 2 --maxDockerSeleniumContainers 6.

**_Errors :_** 
09:55:54.591 [RemoteProxy CleanUpThread for http://172.17.0.8:40307] WARN  o.o.grid.internal.BaseRemoteProxy - session e
xt. key 13a6d501f99ba5fe74a87cb29bbcfa58 has TIMED OUT due to client inactivity and will be released.
09:55:54.600 [RemoteProxy CleanUpThread for http://172.17.0.8:40307] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error rele
asing the resources on timeout for session ext. key 13a6d501f99ba5fe74a87cb29bbcfa58
09:55:54.608 [RemoteProxy CleanUpThread for http://172.17.0.4:40305] WARN  o.o.grid.internal.BaseRemoteProxy - session e
xt. key 2a5a95b27ad4a47fe0acb439aee15625 has TIMED OUT due to client inactivity and will be released.
09:55:54.616 [RemoteProxy CleanUpThread for http://172.17.0.4:40305] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error rele
asing the resources on timeout for session ext. key 2a5a95b27ad4a47fe0acb439aee15625
09:55:54.628 [RemoteProxy CleanUpThread for http://172.17.0.6:40303] WARN  o.o.grid.internal.BaseRemoteProxy - session e
xt. key 23d6d5d4df1dc181d10d446b32505e3f has TIMED OUT due to client inactivity and will be released.
09:55:54.636 [RemoteProxy CleanUpThread for http://172.17.0.6:40303] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error rele
asing the resources on timeout for session ext. key 23d6d5d4df1dc181d10d446b32505e3f
09:55:54.666 [RemoteProxy CleanUpThread for http://172.17.0.3:40304] WARN  o.o.grid.internal.BaseRemoteProxy - session e
xt. key f8188db24c98c64cb091fdfa505e1f59 has TIMED OUT due to client inactivity and will be released.
09:55:54.675 [RemoteProxy CleanUpThread for http://172.17.0.3:40304] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error rele
asing the resources on timeout for session ext. key f8188db24c98c64cb091fdfa505e1f59
09:55:54.707 [RemoteProxy CleanUpThread for http://172.17.0.7:40306] WARN  o.o.grid.internal.BaseRemoteProxy - session e
xt. key 75ba5320a451f933a9d17cecddb494f5 has TIMED OUT due to client inactivity and will be released.
09:55:54.708 [RemoteProxy CleanUpThread for http://172.17.0.5:40308] WARN  o.o.grid.internal.BaseRemoteProxy - session e
xt. key 3bb67fd2a4a4584b90d6fc983a823b1b has TIMED OUT due to client inactivity and will be released.
09:55:54.720 [RemoteProxy CleanUpThread for http://172.17.0.7:40306] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error rele
asing the resources on timeout for session ext. key 75ba5320a451f933a9d17cecddb494f5
09:55:54.727 [RemoteProxy CleanUpThread for http://172.17.0.5:40308] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error rele
asing the resources on timeout for session ext. key 3bb67fd2a4a4584b90d6fc983a823b1b
09:55:59.601 [RemoteProxy CleanUpThread for http://172.17.0.8:40307] WARN  o.o.grid.internal.BaseRemoteProxy - session e
xt. key 13a6d501f99ba5fe74a87cb29bbcfa58 has TIMED OUT due to client inactivity and will be released.
09:55:59.609 [RemoteProxy CleanUpThread for http://172.17.0.8:40307] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error rele
asing the resources on timeout for session ext. key 13a6d501f99ba5fe74a87cb29bbcfa58
09:55:59.617 [RemoteProxy CleanUpThread for http://172.17.0.4:40305] WARN  o.o.grid.internal.BaseRemoteProxy - session e
xt. key 2a5a95b27ad4a47fe0acb439aee15625 has TIMED OUT due to client inactivity and will be released.
09:55:59.625 [RemoteProxy CleanUpThread for http://172.17.0.4:40305] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error rele
asing the resources on timeout for session ext. key 2a5a95b27ad4a47fe0acb439aee15625
09:55:59.637 [RemoteProxy CleanUpThread for http://172.17.0.6:40303] WARN  o.o.grid.internal.BaseRemoteProxy - session e
xt. key 23d6d5d4df1dc181d10d446b32505e3f has TIMED OUT due to client inactivity and will be released.
09:55:59.659 [RemoteProxy CleanUpThread for http://172.17.0.6:40303] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error rele
asing the resources on timeout for session ext. key 23d6d5d4df1dc181d10d446b32505e3f
09:55:59.676 [RemoteProxy CleanUpThread for http://172.17.0.3:40304] WARN  o.o.grid.internal.BaseRemoteProxy - session e
xt. key f8188db24c98c64cb091fdfa505e1f59 has TIMED OUT due to client inactivity and will be released.
09:55:59.685 [RemoteProxy CleanUpThread for http://172.17.0.3:40304] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error rele
asing the resources on timeout for session ext. key f8188db24c98c64cb091fdfa505e1f59
09:55:59.727 [RemoteProxy CleanUpThread for http://172.17.0.7:40306] WARN  o.o.grid.internal.BaseRemoteProxy - session e
xt. key 75ba5320a451f933a9d17cecddb494f5 has TIMED OUT due to client inactivity and will be released.
09:55:59.728 [RemoteProxy CleanUpThread for http://172.17.0.5:40308] WARN  o.o.grid.internal.BaseRemoteProxy - session e
xt. key 3bb67fd2a4a4584b90d6fc983a823b1b has TIMED OUT due to client inactivity and will be released.
09:55:59.735 [RemoteProxy CleanUpThread for http://172.17.0.7:40306] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error rele
asing the resources on timeout for session ext. key 75ba5320a451f933a9d17cecddb494f5
09:55:59.739 [RemoteProxy CleanUpThread for http://172.17.0.5:40308] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error rele
asing the resources on timeout for session ext. key 3bb67fd2a4a4584b90d6fc983a823b1b.

Also I am attaching a image of Zalenium live preview looks like. In screenshot you can see that six idle nodes are there in live preview window and a browser is assigned to it and node is node destroyed
[ZaleniumLivePreview.docx](https://github.com/zalando/zalenium/files/3353875/ZaleniumLivePreview.docx)

**Comment by diemol** [comment]
It is really hard to address the two previous comments since there is not a provided way to reproduce the scenario. If you can provide enough information, please open a new issue.

**Comment by sujata00** [comment]
I am using Zalenium wherein the containers are run through Jenkins jobs. Each job spawns a new container and the testcases are ran on these containers.

If a user aborts the job in between the container is alive and there is no way to figure out how to stop this container (as there are multiple jobs running in parallel).

I have written a python script that kills sessions that are older than 1 day. But it's not of much use.

Is there any new parameter which is supported by Zalenium for determining if the container is in use or not?
